src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,378 @@
+
+            sr_doc = self._sr_cache.pop(sr_uid)
+
+            # POST /api/v1/register/{path}
+            desc_node = self._desc_nodes[desc_uid]
+            data_key = sr_doc["data_key"]
+
+            # Find data shape and machine dtype; dtype_str takes precedence if specified
+            data_desc = dict(desc_node.metadata)["data_keys"][data_key]
+            data_shape = tuple(data_desc["shape"])
+            data_shape = data_shape if data_shape != (1,) else ()
+            data_type = data_desc["dtype"]
+            data_type = DTYPE_LOOKUP[data_type] if data_type in DTYPE_LOOKUP.keys() else data_type
+            data_type = np.dtype(data_desc.get("dtype_str", data_type))
+
+            # Kept for back-compatibility with old StreamResource schema from event_model<1.20.0
+            if ("mimetype" not in sr_doc.keys()) and ("spec" not in sr_doc.keys()):
+                raise RuntimeError("StreamResource document is missing a mimetype or spec")
+            else:
+                sr_doc["mimetype"] = sr_doc.get("mimetype", MIMETYPE_LOOKUP[sr_doc["spec"]])
+            if "parameters" not in sr_doc.keys():
+                sr_doc["parameters"] = sr_doc.pop("resource_kwargs", {})
+            if "uri" not in sr_doc.keys():
+                file_path = sr_doc.pop("root").strip("/") + "/" + sr_doc.pop("resource_path").strip("/")
+                sr_doc["uri"] = "file://localhost/" + file_path
+
+            # Initialise a bluesky handler for the StreamResource
+            handler_class = STREAM_HANDLER_REGISTRY[sr_doc["mimetype"]]
+            handler = handler_class(
+                sr_doc["uri"], data_shape=data_shape, data_type=data_type, **sr_doc["parameters"]
+            )
+
+            sr_node = desc_node["external"].new(
+                key=data_key,
+                structure_family=StructureFamily.array,
+                data_sources=[handler.get_data_source()],
+                metadata=sr_doc,
+                specs=[],
+            )
+
+            self._handlers[sr_uid] = handler
+            self._sr_nodes[sr_uid] = sr_node
+        else:
+            raise RuntimeError(f"Stream Resource {sr_uid} is referenced before being declared.")
+
+        return sr_node, handler
+
+    def stream_datum(self, doc: StreamDatum):
+        # Get the Stream Resource node and handler
+        sr_node, handler = self.get_sr_node(doc["stream_resource"], desc_uid=doc["descriptor"])
+        handler.consume_stream_datum(doc)
+
+        # Update StreamResource node in Tiled; note one-based indexing of data sources.
+        sr_node.refresh()
+        sr_0 = sr_node.data_sources()[0]
+        data_source = handler.get_data_source()
+        data_source.id = sr_0["id"]
+        endpoint = sr_node.uri.replace("/metadata/", "/data_source/", 1)
+        handle_error(
+            sr_node.context.http_client.put(
+                endpoint,
+                content=safe_json_dump({"data_source": data_source}),
+            )
+        ).json()
```

Comment:
Handling for non-Stream Resource/Datum?

Reply:
I'd like to tackle that in a separate PR.